cylc/flow/exceptions.py

```diff
@@ -449,10 +450,12 @@ class NoPlatformsError(PlatformLookupError):
         place: Where the attempt to get the platform failed.
     """
     def __init__(
-        self, identity: str, set_type: str = 'group', place: str = ''
+        self, identity: str, set_type: str = 'group', place: str = '',
+        hosts_consumed: Optional[List[str]] = None
     ):
         self.identity = identity
         self.type = set_type
+        self.hosts_consumed = set(hosts_consumed) if hosts_consumed else set()
```

Comment:
If the `hosts_consumed` field is ever omitted, then the clearing of the bad hosts logic implemented in this PR will not happen.

So this argument really shouldn't be optional as that makes it easy to omit this required information when raising the exception:

```suggestion
    def __init__(
            self,
            identity: str,
            hosts_consumed: Set[str] = None
            set_type: str = 'group',
            place: str = '',
    ):
        self.identity = identity
        self.type = set_type
        self.hosts_consumed = self.hosts_consumed
```